### PR TITLE
Added '2048' command

### LINKED ISSUE #479: 2048 command
image

### PR DESCRIPTION
**Pull request by Acnologla**
Closes #479

### CODE REVIEW

**Comment by fernando457829** [comment]
Please, run `npm run fix` to fix errors from Standard.

**Review by fernando457829 — CHANGES_REQUESTED** [review]
Run `npm run fix` to fix errors from Standard.

**Review by fernando457829 — CHANGES_REQUESTED** [review]
Add command description and usage in `src/locales/en-US/commands.json`.

Something like:
```json
{
  "2048": {
    "commandDescription": "...",
    "commandUsage": "<required_param> [optional_param]"
  }
}
```

File: src/commands/games/2048.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  async run ({ channel, author, t }, boardSize) {
+    const size = boardSize || 4
+    let points = 0
+    const board = this.createBoard(size)
+    let lastPlay = Date.now()
+    for (let i = 0; i < 2; i++) {
+      const empty = this.getEmptyTile(board)
+      board[Math.floor(empty / size)][Math.floor(empty % size)] = 2
+    }
+    const msg = await channel.send(`${this.drawPoints(points)}\n\n${this.drawBoard(board)}`)
```
Comment (moruzerinho6) [inline src/commands/games/2048.js:101]: What if the bot fails to send the message? This isn't even inside a try, it could easily crash.

Comment (pedrofracassi) [reply src/commands/games/2048.js:101]: The bot wouldn't crash, it'd just throw an exception on the console and keep running

File: src/commands/games/2048.js [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+      }
+      const empty = this.getEmptyTile(board)
+      if (empty !== false) {
+        board[Math.floor(empty / size)][Math.floor(empty % size)] = 2
+      }
+      if (board.reduce((acc, val) => acc.concat(val), []).some((tile, i) => tile !== old[i])) {
+        lastPlay = Date.now()
+      }
+      msg.edit(`${this.drawPoints(points)}\n\n${this.drawBoard(board)}`)
+      r.users.remove(author).catch(function () {})
```
Comment (moruzerinho6) [inline src/commands/games/2048.js:147]: What if the user removes itself to intentionally crash the bot?

File: src/commands/games/2048.js (hunk 1 above)
Comment (moruzerinho6) [reply src/commands/games/2048.js:101]: Hm, okay then.

File: src/locales/en-US/commands.json [hunk 3]
```diff
@@ -4,6 +4,10 @@
     "hereIsYour_neko": "Here's your neko!",
     "hereIsYour_lewd": "Here's your lewd neko!"
   },
+  "2048": {
+    "commandDescription": "Play the 2048 game.",
+    "commandUsage": "[board_size]"
```
Comment (metehus) [inline src/locales/en-US/commands.json:9]: maybe change to `[board size]`

File: src/commands/games/2048.js [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    let lastPlay = Date.now()
+    for (let i = 0; i < 2; i++) {
+      const empty = this.getEmptyTile(board)
+      board[Math.floor(empty / size)][Math.floor(empty % size)] = 2
+    }
+    let msg
+    try {
+      msg = await channel.send(`${this.drawPoints(points)}\n\n${this.drawBoard(board)}`)
+    } catch (_) {
+      return
```
Comment (metehus) [inline src/commands/games/2048.js:105]: why is this not handled? at least throw a CommandError

**Review by metehus — CHANGES_REQUESTED** [review]

File: src/commands/games/2048.js (hunk 2 above)
Comment (pedrofracassi) [reply src/commands/games/2048.js:152]: Again, it wouldn't crash the bot, just throw an error.

File: src/commands/games/2048.js [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+      if (empty !== false) {
+        board[Math.floor(empty / size)][Math.floor(empty % size)] = 2
+      }
+      if (board.reduce((acc, val) => acc.concat(val), []).some((tile, i) => tile !== old[i])) {
+        lastPlay = Date.now()
+      }
+      msg.edit(`${this.drawPoints(points)}\n\n${this.drawBoard(board)}`)
+      r.users.remove(author).catch(function () {})
+    })
+  }
```
Comment (pedrofracassi) [inline src/commands/games/2048.js:154]: Please place `run()` function at the top of the class before any other functions.

**Review by pedrofracassi — CHANGES_REQUESTED** [review]
I've been testing the command and comparing to the real game, and it looks like there's one major difference: the command spawns new tiles even when there are no possible moves, whereas the original game does not.

You might want to check if anything has moved before spawning a new tile.

File: src/commands/games/2048.js (hunk 5 above)
Comment (moruzerinho6) [reply src/commands/games/2048.js]: He fixed in 9f7952f

Comment (pedrofracassi) [reply src/commands/games/2048.js]: Cool

**Review by moruzerinho6 — DISMISSED** [review]
Works on my machine™

File: src/commands/games/2048.js [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    for (let i = 0; i < board.length; i++) {
+      board[i] = rotatedBoardArray[i]
+    }
+  }
+  drawBoard (board) {
+    let text = ''
+    for (const row of board) {
+      for (const tile of row) {
+        const stringTile = tile.toString()
+        text += this.getEmoji(stringTile.length === 1 ? stringTile + '_' : stringTile)
```
Comment (pedrofracassi) [inline src/commands/games/2048.js:121]: Just to match how we've saved the emoji:
```suggestion
        text += this.getEmoji(`${stringTile}_`)
```

**Review by pedrofracassi — DISMISSED** [review]
Looking good!

File: src/commands/games/2048.js [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+      for (let j = 0; j < board.length; j++) {
+        if (board[i][j] === 0) {
+          empty.push(j + (i * board.length))
+        }
+      }
+    }
+    if (empty.length === 0) {
+      return false
+    }
+    return empty[Math.floor(Math.random() * empty.length)]
```
Comment (metehus) [inline src/commands/games/2048.js:153]:
```suggestion
	return empty.length ? empty[Math.floor(Math.random() * empty.length)] : false
```

**Review by metehus — DISMISSED** [review]

File: src/commands/games/2048.js (hunk 7 above)
Comment (Acnologla) [reply src/commands/games/2048.js]: Fixed in [d402ff9](https://github.com/SwitchbladeBot/switchblade/pull/1153/commits/d402ff9882ee0ebc51817c41147022e5e53f0a5a)

**Review by pedrofracassi — APPROVED** [review]

**Comment by pedrofracassi** [comment]
Played a full match and it didn't mark it as ended once no more moves were possible. Is this intended?

Discord_6RhMSOZPC1

**Review by fernando457829 — APPROVED** [review]

**Pull request merged**